lib/signet/oauth_2/client.rb

```diff
@@ -1020,10 +1046,13 @@ def fetch_access_token options = {}
           content_type = response.header[:content_type]
         end
 
-        return ::Signet::OAuth2.parse_credentials body, content_type if status == 200
-
         message = "  Server message:\n#{response.body.to_s.strip}" unless body.to_s.strip.empty?
-        if [400, 401, 403].include? status
+
+        if status == 200
+          parsed_response = ::Signet::OAuth2.parse_credentials body, content_type
+          parsed_response["granted_scopes"] = parsed_response.delete("scope") if parsed_response
```

Comment:
Can you explain this line to me please? I am not familiar with Ruby nor this library.

Reply:
We already have a field called `scope` which represents the requested scope. The `update_token` method updates the field `scope` if the `options` hash passed to the method contains a key `scope`. I need to make sure that we're not mixing the requested scope and granted scopes. So, I create a separate key called `granted_scopes` ( and delete the `scope` key so that it doesn't get updated for requested scope ) and pass it to update_token method.

Reply:
IIUC, "scope" field was already existing, if options had it, then the field is set. Now, that functionality is being removed?

I am trying to understand why both "Scopes" and "granted_scopes" cannot co-exist.

Reply:
```
IIUC, "scope" field was already existing, if options had it, then the field is set. Now, that functionality is being removed?
```
The scope field here represents granted_scopes while the already existing scope field represents requested scopes. We need to differentiate the two and hence this deletion .

Reply:
Let me know if it isn't clear. We can discuss it offline